pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/BalanceNumSegmentAssignmentStrategy.java

```diff
@@ -0,0 +1,121 @@
+
+    // Assign the segment to the instance with the least segments, or the smallest id if there is a tie
+    int numInstances = numSegmentsAssigned.length;
+    PriorityQueue<Pairs.IntPair> heap = new PriorityQueue<>(numInstances, Pairs.intPairComparator());
+    for (int instanceId = 0; instanceId < numInstances; instanceId++) {
+      heap.add(new Pairs.IntPair(numSegmentsAssigned[instanceId], instanceId));
+    }
+    List<String> instancesAssigned = new ArrayList<>(_replication);
+    for (int i = 0; i < _replication; i++) {
+      instancesAssigned.add(instances.get(heap.remove().getRight()));
+    }
+
+    LOGGER.info("Assigned segment: {} to instances: {} for table: {}", segmentName, instancesAssigned,
+        _tableNameWithType);
+    return instancesAssigned;
+  }
+
+  @Override
+  public Map<String, Map<String, String>> rebalanceTable(Map<String, Map<String, String>> currentAssignment,
+      InstancePartitions instancePartitions) {
+    List<String> instances = getInstances(instancePartitions);
+
+    // Use Helix AutoRebalanceStrategy to rebalance the table
+    LinkedHashMap<String, Integer> states = new LinkedHashMap<>();
+    states.put(SegmentOnlineOfflineStateModel.ONLINE, _replication);
+    AutoRebalanceStrategy autoRebalanceStrategy =
+        new AutoRebalanceStrategy(_tableNameWithType, new ArrayList<>(currentAssignment.keySet()), states);
+    // Make a copy of the current assignment because this step might change the passed in assignment
+    Map<String, Map<String, String>> currentAssignmentCopy = new TreeMap<>();
+    for (Map.Entry<String, Map<String, String>> entry : currentAssignment.entrySet()) {
+      String segmentName = entry.getKey();
+      Map<String, String> instanceStateMap = entry.getValue();
+      currentAssignmentCopy.put(segmentName, new TreeMap<>(instanceStateMap));
+    }
+    Map<String, Map<String, String>> newAssignment =
+        autoRebalanceStrategy.computePartitionAssignment(instances, instances, currentAssignmentCopy, null)
+            .getMapFields();
+
+    LOGGER.info(
+        "Rebalanced {} segments with {} replicas to instances: {} for table: {}, number of segments to be moved to each instance: {}",
+        currentAssignment.size(), _replication, instances, _tableNameWithType,
+        SegmentAssignmentUtils.getNumSegmentsToBeMoved(currentAssignment, newAssignment));
+    return newAssignment;
+  }
+
+  private List<String> getInstances(InstancePartitions instancePartitions) {
+    Preconditions.checkArgument(instancePartitions.getNumPartitions() == 1,
+        "The instance partitions should contain only 1 partition");
+    Preconditions.checkArgument(instancePartitions.getNumReplicas() == 1,
+        "The instance partitions should contain only 1 replica");
+    List<String> instances = instancePartitions.getInstances(0, 0);
+    Preconditions.checkState(instances.size() >= _replication,
```

Comment:
it will be useful to indicate how we arrived at less number of instances. Since instance assignment is now multiple steps, just returning the end result will leave an administrator confused as to what they did wrong.

Reply:
That will be handled in the instance assignment step (in the upcoming pr). This check is for cases where people modify table replication without re-calculate instance partitions.